src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -15,65 +15,182 @@
+                InputStreamReader inputStreamReader = new InputStreamReader(fileInputStream, StandardCharsets.UTF_8);
+                BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {
+
+            // Seek to the start of the first complete line in the chunk
+            if (start != 0) {
+                file.seek(start);
+            }
+
+            long processedUpTo = start;
 
-//        long before = System.currentTimeMillis();
+            Map<String, Measurement> measurements = new HashMap<>(700); // kenough room.
 
-        Map<String, Measurement> resultMap = Files.lines(Path.of(FILE)).parallel()
-                .map(record -> {
-                    // Map to <String,double>
-                    int pivot = record.indexOf(";");
-                    String key = record.substring(0, pivot);
-                    double measured = Double.parseDouble(record.substring(pivot + 1));
-                    return new AbstractMap.SimpleEntry<>(key, measured);
-                })
-                .collect(Collectors.toConcurrentMap(
-                        // Combine/reduce:
-                        AbstractMap.SimpleEntry::getKey,
-                        entry -> new Measurement(entry.getValue()),
-                        Measurement::combineWith));
+            String line;
+            while (processedUpTo < end && (line = bufferedReader.readLine()) != null) {
 
-        System.out.print("{");
-        System.out.print(
-                resultMap.entrySet().stream().sorted(Map.Entry.comparingByKey()).map(Object::toString).collect(Collectors.joining(", ")));
-        System.out.println("}");
+                processedUpTo += line.length() + 1;
 
-//        System.out.println("Took: " + (System.currentTimeMillis() - before));
+                // Process the line
+                int pivot = line.indexOf(";");
+                String key = line.substring(0, pivot);
+                int measured = branchlessParseInt(line.substring(pivot + 1));
+
+                // Tried merge() and compute() but this seems to be fastest
+                Measurement measurement = measurements.computeIfAbsent(key, (k) -> new Measurement());
+                measurement.updateWith(measured);
+            }
+            return measurements;
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
 
+    // Specialized base-10 parser, skips - and .
+    private static int branchlessParseInt(final String input) {
+        // Branchless parser, goes from:
+        // "-1.2" to -12
+        // "40.1" to 401
+        // etc.
+
+        // Magic.
+        int off = ~(input.charAt(0) >> 4) & 1;
+        int neg = 1 - off - off;
+        int has4 = ((input.length() - off) >> 2) & 1;
+        return neg * ((input.charAt(off) - '0') * 100 * has4 + (input.charAt(off + has4) - '0') * 10 + (input.charAt(2 + off + has4) - '0'));
+    }
```

Comment:
Don't ask me about this code, its magic.

It takes these numbers in the following formats:
-X.X
-XX.X
X.X
XX.X

and turns them into integers XXX (10x the original amount) without any branches.

Turns out chatAt() is faster in this case than getting the array and referencing that.

Reply:
I think I've tried that, it is something the compiler instantly picks up anyway, it knows these variables are final and I didn't see a performance difference. They probably get smooshed into a couple/single instructions anyway since they have such a limited scope.